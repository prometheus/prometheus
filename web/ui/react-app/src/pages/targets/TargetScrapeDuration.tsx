import React, { FC, Fragment, useState } from 'react';
import { Tooltip } from 'reactstrap';
import 'css.escape';
import { humanizeDuration } from '../../utils';

interface Labels {
  [key: string]: string;
}

export interface TargetScrapeDurationProps {
  duration: number;
  labels: Labels;
  idx: number;
  scrapePool: string;
}

const TargetScrapeDuration: FC<TargetScrapeDurationProps> = ({ duration, labels, idx, scrapePool }) => {
  const [scrapeTooltipOpen, setScrapeTooltipOpen] = useState<boolean>(false);
  const id = `scrape-duration-${scrapePool}-${idx}`;

  return (
    <>
      <div id={id} className="scrape-duration-container">
        {humanizeDuration(duration * 1000)}
      </div>
      <Tooltip
        isOpen={scrapeTooltipOpen}
        toggle={() => setScrapeTooltipOpen(!scrapeTooltipOpen)}
        target={CSS.escape(id)}
        style={{ maxWidth: 'none', textAlign: 'left' }}
      >
        {labels['__scrape_interval__'] ? (
          <Fragment>
            <span>Interval: {labels['__scrape_interval__']}</span>
            <br />
          </Fragment>
        ) : null}
        {labels['__scrape_timeout__'] ? (
          <Fragment>
            <span>Timeout: {labels['__scrape_timeout__']}</span>
          </Fragment>
        ) : null}
      </Tooltip>
    </>
  );
};

export default TargetScrapeDuration;
