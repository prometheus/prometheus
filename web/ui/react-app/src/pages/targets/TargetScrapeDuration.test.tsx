import * as React from 'react';
import { shallow } from 'enzyme';
import TargetScrapeDuration from './TargetScrapeDuration';
import { Tooltip } from 'reactstrap';
import toJson from 'enzyme-to-json';

describe('targetScrapeDuration', () => {
  const defaultProps = {
    duration: 0.008,
    labels: {
      __address__: 'localhost:9100',
      __metrics_path__: '/metrics',
      __scheme__: 'http',
      __scrape_interval__: '15s',
      __scrape_timeout: '10s',
      job: 'node_exporter',
    },
    idx: 1,
    scrapePool: 'cortex/node-exporter_group/0',
  };
  const targetScrapeDuration = shallow(<TargetScrapeDuration {...defaultProps} />);

  it('renders duration correctly', () => {
    const duration = targetScrapeDuration.find('.scrape-duration-container');
    expect(duration.text()).toBe('8.000ms');
  });

  it('renders a tooltip for scrape interval and timeout', () => {
    const tooltip = targetScrapeDuration.find(Tooltip);
    expect(tooltip).toHaveLength(1);
    expect(tooltip.prop('isOpen')).toBe(false);
    expect(tooltip.prop('target')).toEqual('scrape-duration-cortex\\/node-exporter_group\\/0-1');
  });

  it('renders scrape interval and timeout', () => {
    expect(toJson(targetScrapeDuration)).toMatchSnapshot();
  });
});
