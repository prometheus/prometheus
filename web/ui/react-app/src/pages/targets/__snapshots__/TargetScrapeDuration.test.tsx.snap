// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`targetScrapeDuration renders scrape interval and timeout 1`] = `
<Fragment>
  <div
    className="scrape-duration-container"
    id="scrape-duration-cortex/node-exporter_group/0-1"
  >
    8.000ms
  </div>
  <Tooltip
    autohide={true}
    isOpen={false}
    placement="top"
    placementPrefix="bs-tooltip"
    style={
      Object {
        "maxWidth": "none",
        "textAlign": "left",
      }
    }
    target="scrape-duration-cortex\\\\/node-exporter_group\\\\/0-1"
    toggle={[Function]}
    trigger="hover focus"
  >
    <span>
      Interval: 
      15s
    </span>
    <br />
  </Tooltip>
</Fragment>
`;
